base/unicode/utf8.jl

```diff
@@ -312,6 +312,11 @@ function convert(::Type{UTF8String}, a::Vector{UInt8}, invalids_as::AbstractStri
 end
 convert(::Type{UTF8String}, s::AbstractString) = utf8(bytestring(s))
 
+function convert{T<:Union{UInt16, UInt32}}(::Type{UTF8String}, a::Vector{T}, len=length(a)-1)
+    encode_to_utf8(T, a, len)
```

Comment:
I think you are passing the wrong `len` here.   The third argument of `encode_to_utf8` is the length of the resulting UTF-8 string, in bytes.

See the `convert(::Type{UTF8String}, str::UTF16String)` function in `utf16.jl`.  The best thing would probably be to refactor that function into:

```jl
function convert(::Type{UTF8String}, dat::Vector{UInt16), len=length(dat))
    len < 1 && return empty_utf8
    # get number of bytes to allocate
    len, flags, num4byte, num3byte, num2byte = unsafe_checkstring(dat, 1, len)
    flags == 0 && @inbounds return UTF8String(copy!(Vector{UInt8}(len), 1, dat, 1, len))
    return encode_to_utf8(UInt16, dat, len + num2byte + num3byte*2 + num4byte*3)
end

function convert(T::Type{UTF8String}, str::UTF16String)
    return convert(T, str.data, length(str.data) - 1)
end
```
and similarly for UTF-32.